src/fundamental/chid-fundamental.c

```diff
@@ -0,0 +1,118 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+
+/*
+ * Based on Nikita Travkin's dtbloader implementation.
+ * Copyright (c) 2024 Nikita Travkin <nikita@trvn.ru>
+ *
+ * https://github.com/TravMurav/dtbloader/blob/main/src/chid.c
+ */
+
+/*
+ * Based on Linaro dtbloader implementation.
+ * Copyright (c) 2019, Linaro. All rights reserved.
+ *
+ * https://github.com/aarch64-laptops/edk2/blob/dtbloader-app/EmbeddedPkg/Application/ConfigTableLoader/CHID.c
+ */
+
+#if SD_BOOT
+#  include "efi-string.h"
+#  include "util.h"
+#else
+#  include <byteswap.h>
+#  include <string.h>
+#  include <uchar.h>
+#  include <utf8.h>
+#define strsize16(str) ((char16_strlen(str) + 1) * sizeof(char16_t))
+#endif
+
+#include "macro-fundamental.h"
+#include "memory-util-fundamental.h"
+#include "chid-fundamental.h"
+#include "sha1-fundamental.h"
+
+static void get_chid(const char16_t *const smbios_fields[static _CHID_SMBIOS_FIELDS_MAX], uint32_t mask, EFI_GUID *ret_chid) {
+        assert(mask);
+        assert(ret_chid);
+        uint8_t hash[SHA1_DIGEST_SIZE] = {};
+        const EFI_GUID namespace = { 0x12d8ff70, 0x7f4c, 0x7d4c, {} }; /* Swapped to BE */
```

Comment:
even if this will result in the right values, it does feel really weird to use hex here and specifc signed values that will be converted to unsigned anyway. consider suffixing these values with U, or even better use UINT32_C(), UINT16_C() on the values to remove any doubt